### ISSUE TITLE
Question Modbus Register Request String

### ISSUE THREAD

**Issue by tmycgn** [TheChatty/SmartESPatMax#4]
I slightly struggle with how you have specified the holding registers of the Trovis in the Tasmota Smart Meter Interface syntax ([https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-metrics](https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-metrics))...

Your example contains:

```
2,F7031CSSss@i0:10,Außentemp.,°C,Temp_Outside,1
2,F7031Cx06SSss@i0:10,Vorlauftemp.,°C,Temp_Flow,1
2,F7031Cx14SSss@i0:10,Rücklauftemp.,°C,Temp_Return,1
```

I use the presumably well-known list from Tom-Bom-badil as a reference:
[https://github.com/Tom-Bom-badil/samson_trovis_557x/blob/ffaadbedb47476c05111ccc42d41ae8655a9a419/SmartHomeNG/shNG-plugins/trovis557x/_register.py](https://github.com/Tom-Bom-badil/samson_trovis_557x/blob/ffaadbedb47476c05111ccc42d41ae8655a9a419/SmartHomeNG/shNG-plugins/trovis557x/_register.py)

The Außentemperatur here has register 40010, the Vorlauftemp. 40013, the Rücklauftemp. 40017.
I understand that the **F7** is your modbus slave ID, which **03** defines the readout of the holding register. Unfortunately, I don't understand the rest.

How do you get from **CSSss** to 40010 and from **Cx06SSss** to 4013?

I also know [https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-metrics](https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-metrics) but I still can't manage it. 🙈

I would really appreciate your answer and clarification, as I would like to add more registers to my configuration. 🎉

**Comment by TheChatty** [comment]
Yes, finding out about this matter was a long trail of trial and error. And If you want to read out multiple meters with a single ESP8266 you have to take care of the order of meters in your script (changing them leads to some meter(s) not working). But if you have only one meter this is not a problem.

The meter definition shows my example issues three read operations (first one is: `...,rF7030009000E,...`). The first one will read all registers from 9 to 22 (zero-based, numOfRegs: `0x0E`, 2 bytes (signed word: `SSss`) per reg) into variable `i0`.

`F7 03 1C` matches the beginning of the response. First register (#9) comes directly after it, so `SS ss`. The temperature value is multiplied with 10 - so we have to divide by 10 (`i0:10`) to gain one decimal place.
`F7 03 1C` matches the beginning of the response. Second register (#13) must skip 3 regs / 6 bytes --> `x06`. It's also a signed word (`SSss`). It's also multiplied with 10 and needs to be divided...

**Comment by tmycgn** [comment]
Vielen Dank @TheChatty !!

Hab gesehen dass der andere Issue auch auf deutsch war, daher Switch der Sprache. 😉

Das mit den Blöcken und Variablen hatte ich komplett nicht gesehen und somit nicht verstanden, jetzt langsam wird es mir klar. 🙈 Vielleicht kannst du mir noch 4 kurze Fragen beantworten:

- Wann wird das `r` davor gesetzt bei der Blockdefinition, steht das für Range? `rF7030009000E`
- Warum hast du die Register-Range 9-14 für den i0-Block gewählt? Weil die _Speichertemperatur_ (`F7031Cx26SSss`) mir 13 Registern Skip der letzte Wert in dem Block ist?
- Wenn `F7 03 1C` eine Antwort bezeichnet, warum beginnt der zweite Block (i1) dann bei Dir mit `F7 03 04` ? Bzw. was ist die `1C `oder die `04`?
- Verständnis-Prüfung somit: Adresse 13 (VorlauftempVF2) wäre dann wie folgt: Skippen von 4 regs / 8 bytes --> x08.
Somit: `F7031Cx08SSss`

Hintergrund der Frage, ich habe eigentlich wie Du die Trovis 5573, bei mir kommt aber bei den Werten nichts raus.
Vermutlich muss ich andere Register nehmen.

**Comment by TheChatty** [comment]
I'll stick to English so the "world" may read with us...

`r` stands for "Raw binary coding (any binary telegram)" (see `<type>`).

`0x1C` = 2x `0x0E` because there are two bytes per register. So it denotes the length of the response.

You can either issue a single read for every needed register or read them as a block... this requires that their register number is close to each other. Every ModBus telegram is restricted to 255 bytes length (including header & footer).

It might be that I chose the read block a little too big  There are a few ModBus description sites out there if you want to dig deeper into the protocol. I'm happy I'm past that point.

**Comment by tmycgn** [comment]
I made good progress yesterday and was able to add more registers with your help. 🎉

There's only one more problem: I can retrieve holding registers via the `03`, but no coils such as the **binary outputs like UP1, ZP, SLP (registers 56-60)** Theoretically you can read out a coil via the `01` which only contains 1 bit, but I can't do it with Tasmota. Have you also run into this and do you have a solution or approach? Unfortunately, I cannot find any information on this at https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-definition.
